### ISSUE TITLE
Discussion on ESP32 and W5500

### ISSUE THREAD

**Issue by HWEDK** [KaeLL/ESP32-W5100#2]
I was hoping more to ask a few questions that would really help save development time as well as to add some insight into the W5100 versus W5500.

I am trying to get the W5500 to work with the ESP-IDF Asynchronous webserver functions, but discovered as I'm sure you did that the TCP/IP nature of the W5500 is not compatible directly with the lwIP ESP32 protocol. 

1) Are you mainly implementing the MACRAW function of the W5100/W5500 to bypass the TCP/IP engine, or are there other pieces to consider? (I know the W5500 is faster, has more sockets, etc. but not exactly what may cause issues with lwIP MACRAW)

2) Do you think it would be relatively doable to port over what you discovered to W5500?

3) Will the W5100 implementation for ESP32, that you wrote, allow use for all the lwIP functions in the IDF?

**Comment by KaeLL** [comment]
1. Not exactly mainly, but rather "only". So yes, in regards to this, my driver and the W5500 one merged on IDF recently are basically the same. And I think you mean [ESP-NETIF instead of LwIP](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/network/esp_netif.html#esp-netif-architecture)? Either way, I don't know of any issues between the two. The fundamental problem in this scenario is that both Wiznet chips have an EthMAC interface in hardware, making them a sub-optimal choice for integration with ESP32 because of the redundancy, since ESP32 also has one.

2. Yes, but there's no need, as the driver for the W5500 [was merged into IDF](https://github.com/espressif/esp-idf/tree/master/components/esp_eth/src) and is available in release since [v4.3-beta1](https://github.com/espressif/esp-idf/releases/tag/v4.3-beta1). I take from your use of the async webserver that you're using the Arduino framework, and I have no idea how long it takes for things to be available in it.

3. If you really mean LwIP here, the answer is yes, as LwIP is a Network-layer-and-above library and it doesn't concern itself with anything below it. As for ESP-NETIF, the short answer is yes. The longer one is also yes, but you may not have as much control over the PHY interface as you'd have wished especially in W5100's case, assuming that's relevant for your application (which I'm more than sure it isn't).

**Comment by HWEDK** [comment]
Okay, thank you. I should note that much of this is new territory for me. I did not know about ESP-NETIF, and had only been looking at implementation that involves LwIP. You are correct in that I am not concerned much with deep lower level control. I am more concerned with using multiple protocols with Asynchronous functionality that in part involve Ethernet as an option. 

So, the W5500 being integrated as far as I can see in the commits to the newer ESP-IDF seems to include the MACRAW functions and board setup. I am using Arduino IDE, but if the new Ethernet libraries will work with ESP32 decently then it may be worth migrating to another IDE--difficult to do offline. 

If you have used another implementation that would allow for HTTP communication through a basic fileserver like the FatFS/SPIFFS etc smoothly--and that would work with WiFi--I'd be curious to try that. Otherwise, thank you for the feedback and pointing me to ESP-NETIF as well as the new v4.3-beta1

**Comment by KaeLL** [comment]
> If you have used another implementation that would allow for HTTP communication through a basic fileserver like the FatFS/SPIFFS etc smoothly--and that would work with WiFi--I'd be curious to try that.

ESP-IDF has all that ([1](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/protocols/esp_http_server.html), [2](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/storage/fatfs.html), [3](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/storage/spiffs.html)) and examples that you can use as a starting point.
One of the reasons I'm not a fan of the Arduino ecosystem for the ESP32 is because it overlaps too much with ESP-IDF. It's built upon the assumption that your MCU is simple and that there's little-to-no vendor support, both of which are false in the case of the ESP32. Since it's a simple wrapper over core/basic functionality, you either will end up using ESP-IDF libraries, in which case the Arduino code will consist only of startup boilerplate, rendering the Arduino part of your firmware pointless, or you'll use 3rd-party libs in true Arduino fashion, missing out on what's already provided by ESP-IDF.

Feel free to close the issue if/when you have no more questions or comments.

**Comment by HWEDK** [comment]
I realize I'm moving a bit away from the main topic--I majorly appreciate the insight--but if I may go digress once more...

I am beginning to understand that paradigm of Arduino uselessness for deeper and expanding functionality. My largest issue is then getting Eclipse or similar setup with the IDF offline, without being about to source dependencies using the update python/IDE tools directly--for reasons. 

I guess the only reason I'm still pushing forward is that I got the filesystem, Wi-Fi, and Ethernet to run HTTP protocols smoothly with minimal overhead in Arduino simultaneously. If that can be rebuilt relatively simply using the proper ESP-IDF in a way that uses relatively simple asynchronous HTTP commands using the fileserver, then I definitely would prefer that. I'm worried that managing the partitions and serving data from HTTP requests between WiFi/Ethernet would increase complexity beyond available timeline. 

My final question would be, do you think the dependencies could be relatively easily setup offline for Eclipse/Visual Studio or similar? Then, is there enough info on the filesystem to use the asynch server similar to how the below randomnerd example does without building a major amount of firmware changes? 

https://randomnerdtutorials.com/esp32-web-server-spiffs-spi-flash-file-system/

**Comment by KaeLL** [comment]
> My largest issue is then getting Eclipse or similar setup with the IDF offline, without being about to source dependencies using the update python/IDE tools directly--for reasons.
...
do you think the dependencies could be relatively easily setup offline for Eclipse/Visual Studio or similar?

Not sure how that works for Arduino, but for ESP-IDF, you need to run the [install](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/get-started/#step-3-set-up-the-tools) script only once, and it downloads all tools needed (and installs the python dependencies). After that, you can go offline until you update ESP-IDF, should you choose to do so.

> If that can be rebuilt relatively simply using the proper ESP-IDF in a way that uses relatively simple asynchronous HTTP commands using the fileserver, then I definitely would prefer that. I'm worried that managing the partitions and serving data from HTTP requests between WiFi/Ethernet would increase complexity beyond available timeline.
...
is there enough info on the filesystem to use the asynch server similar to how the below randomnerd example does without building a major amount of firmware changes?

Than you should probably stick with what you have, as it's possible you'd have to code the integration between these components yourself. But that's if you can't find someone else's code that already does that, which I have no idea about the odds, as I've never searched for it.

**Comment by KaeLL** [comment]
Closing it. Feel free to reopen if needed.
